sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -319,14 +362,19 @@ final class DataFrameWriter private[sql](df: DataFrame) {
         checkpointPath.toUri.toString
       }
 
-      val sink = new MemorySink(df.schema)
+      if (!Seq(OutputMode.Append, OutputMode.Complete).contains(outputMode)) {
```

Comment:
nit: maybe move this logic to the constructor of `MemorySink` so that we can make sure no place will pass a wrong OutputMode to MemorySink.